sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionPump.cs

```diff
@@ -0,0 +1,220 @@
+
+                        RunningTask = RunAsync(RunningTaskTokenSource.Token);
+                    }
+                }
+                finally
+                {
+                    RunningTaskSemaphore.Release();
+                }
+            }
+        }
+
+        /// <summary>
+        ///   Stops the partition pump.  In case it hasn't been started, nothing happens.
+        /// </summary>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        public Task StopAsync() => StopAsync(PartitionProcessorCloseReason.Shutdown);
+
+        /// <summary>
+        ///   Stops the partition pump.  In case it hasn't been started, nothing happens.
+        /// </summary>
+        ///
+        /// <param name="reason">The reason why the partition pump is being closed.</param>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        private async Task StopAsync(PartitionProcessorCloseReason reason)
+        {
+            if (RunningTask != null)
+            {
+                await RunningTaskSemaphore.WaitAsync().ConfigureAwait(false);
+
+                try
+                {
+                    if (RunningTask != null)
+                    {
+                        RunningTaskTokenSource.Cancel();
+                        RunningTaskTokenSource = null;
+
+                        await RunningTask.ConfigureAwait(false);
+                        RunningTask = null;
+
+                        await InnerConsumer.CloseAsync().ConfigureAwait(false);
+                        InnerConsumer = null;
+
+                        await PartitionProcessor.CloseAsync(reason).ConfigureAwait(false);
+                    }
+                }
+                finally
+                {
+                    RunningTaskSemaphore.Release();
+                }
+            }
+        }
+
+        /// <summary>
+        ///   The main loop of a partition pump.  It receives events from the Azure Event Hubs service
+        ///   and delegates their processing to the inner partition processor.
+        /// </summary>
+        ///
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> instance to signal the request to cancel the operation.</param>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        private async Task RunAsync(CancellationToken cancellationToken)
+        {
+            var retryPolicy = new BasicRetryPolicy(new RetryOptions());
+
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                IEnumerable<EventData> receivedEvents = null;
+
+                try
+                {
+                    receivedEvents = await InnerConsumer.ReceiveAsync(Options.MaximumMessageCount, Options.MaximumReceiveWaitTime).ConfigureAwait(false);
```

Comment:
You probably want to pass the cancellation token along to the `ReceiveAsync` method and potentially the `ProcessError` and `ProcessEvents` if they actually take a token... and if not, should they?